### PR TITLE
PE-4454: Add Promo Code Field to Turbo Modal

### PR DESCRIPTION
**Pull request by matibat**
--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/0ivtcq3m4nu38

### CODE REVIEW

**Comment by matibat** [comment]
@matibat, the tests are not passing!

**Review by thiagocarvalhodev — COMMENTED** [review]
@matibat , the promo field doesn't apply the promo, even though the text field shows the green border indicating the code entered is valid, but no promotion or any other feedback is shown

https://github.com/ardriveapp/ardrive-web/assets/32248947/99443d0f-b544-4fa5-9950-8ffd8166a7b5

**Review by thiagocarvalhodev — CHANGES_REQUESTED** [review]
Left a couple of comments and some required changes.

File: lib/turbo/topup/blocs/payment_review/payment_review_state.dart [hunk 1]
```diff
@@ -42,13 +45,15 @@ class PaymentReviewPaymentModelLoaded extends PaymentReviewState {
     String? subTotal,
     String? credits,
     DateTime? quoteExpirationDate,
-    PaymentUserInformation? userInformation,
+    // PaymentUserInformation? userInformation,
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/turbo/topup/blocs/payment_review/payment_review_state.dart:48]: Ahh yes. I was pretty sure we were not using that piece of data. IMO we can remove that.

**Comment by matibat** [comment]
RE:
> @matibat , the promo field doesn't apply the promo, even though the text field shows the green border indicating the code entered is valid, but no promotion or any other feedback is shown

You have to click the Apply button, although it seems it's not visible on the light theme.

Also, the only three valid Codes I've enabled are `TURBO`, `ARDRIVE`, and `MATI`.

**Comment by matibat** [comment]
Added an extra code: `ERROR` that will simulate a network error.

File: lib/turbo/turbo.dart [hunk 2]
```diff
@@ -186,6 +192,43 @@ class Turbo extends Disposable {
...
+  //       // setState(() {
+  //       //   _promoCode = '';
+  //       //   _promoCodeInvalid = true;
+  //       //   _promoCodeController.clear();
+  //       // });
+  //     }
+  //   }
+  // }
+
+  Future<double?> getPromoDiscountFactor(String promoCode) async {
```
Comment (thiagocarvalhodev) [inline lib/turbo/turbo.dart:217]: The Turbo class should use an abstraction for the logic of the promo code and implement a method using an abstraction. We can see similar examples on the other methods. 

What I suggest is to implement this method in a new class to retrieve the promo code and that also validates the code, example: `TurboPromoDiscountValidator`, that will get a promo code, fetch the API and validate if it's valid or not depending on the service response.

We can also implement a method on the `PaymentService` class to mock the http call, and use this class on the `TurboPromoDiscountValidator`, so, when the backend is done, we can just replace the call on the payment service and we won't change anything else in the code.

Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart:217]: With that, we can also implement the unit tests validating a mocked response from the `PaymentService`

File: lib/turbo/turbo.dart [hunk 3]
```diff
@@ -186,6 +192,43 @@ class Turbo extends Disposable {
 
     logger.d('Turbo disposed');
   }
+
+  // Future<double?> _getPromoDiscountFactor(String promoCode) async {
```
Comment (thiagocarvalhodev) [inline lib/turbo/turbo.dart:196]: Commented code

File: test/turbo/blocs/payment_form/payment_form_bloc_test.dart [hunk 4]
```diff
@@ -16,7 +16,11 @@ void main() {
     setUp(() {
       mockTurbo = MockTurbo();
       initialPriceEstimate = PriceEstimate(
-          credits: BigInt.from(10), priceInCurrency: 10, estimatedStorage: 1);
+        credits: BigInt.from(10),
```
Comment (thiagocarvalhodev) [inline test/turbo/blocs/payment_form/payment_form_bloc_test.dart:19]: We are not testing the case when the promo discount is different than 0.
We need to test:
 - promo discount different from 0
 - promo discount has an error

**Review by thiagocarvalhodev — COMMENTED** [review]
We will need just more one refactor on the `Turbo` class, and also to update the unit tests on the PaymentFormBloc

**Comment by thiagocarvalhodev** [comment]
Now the feature is looking great! Thanks for the fix on Theme, @matibat

File: lib/turbo/turbo.dart (hunk 2 above)
Comment (matibat) [reply lib/turbo/turbo.dart]: Is this what you meant? - 5912114a34c596c110d74bf223d66bd60dd65cef

File: lib/turbo/topup/blocs/payment_review/payment_review_state.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (matibat) [reply lib/turbo/topup/blocs/payment_review/payment_review_state.dart]: 26aa7a40c25351c08ec4d70cc89cf88290d07a96

File: lib/turbo/topup/views/topup_payment_form.dart [hunk 5]
```diff
@@ -574,6 +596,178 @@ class TurboPaymentFormViewState extends State<TurboPaymentFormView> {
...
+    });
+    await Future.delayed(const Duration(seconds: 2));
+    setState(() {
+      _promoCodeFetching = false;
+    });
+
+    final isValid = validCodes.keys.contains(textInPromoCode);
+
+    return isValid ? validCodes[textInPromoCode] : null;
+  }
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/turbo/topup/views/topup_payment_form.dart]: Applied

File: lib/turbo/turbo.dart (hunk 3 above)
Comment (matibat) [reply lib/turbo/turbo.dart]: Removed

File: test/turbo/blocs/payment_form/payment_form_bloc_test.dart (hunk 4 above)
Comment (matibat) [reply test/turbo/blocs/payment_form/payment_form_bloc_test.dart:19]: The bloc won't behave differently for a different promoDiscountFactor. So the zero value is not distinct (in behavior) from one.

Also, there's no error case for this bloc.

Comment (thiagocarvalhodev) [reply test/turbo/blocs/payment_form/payment_form_bloc_test.dart:19]: >Also, there's no error case for this bloc.

I mean the test for emitting this state:

```dart
emit(
        PaymentFormLoaded(
          state.priceEstimate,
          _expirationTimeInSeconds(turbo.maxQuoteExpirationDate),
          (state as PaymentFormLoaded).supportedCountries,
          isPromoCodeInvalid: false,
          isFetchingPromoCode: false,
          errorFetchingPromoCode: true,
        ),
      );

File: lib/turbo/turbo.dart (hunk 2 above)
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart]: Exactly! We are now just missing the unit tests

File: test/turbo/blocs/payment_form/payment_form_bloc_test.dart (hunk 4 above)
Comment (matibat) [reply test/turbo/blocs/payment_form/payment_form_bloc_test.dart:19]: Makes the class be testable by injecting a method to make it be deterministic - d228ee900704c19d7bdf22a691f6d9ed08973d0d
Unit tests those cases - 2428e7394be807a3685c3819320be238072c1189

**Review by thiagocarvalhodev — APPROVED** [review]
Tested and it's working nicely! I just noticed that we can improve the UX by a simple change. @matibat , I think it would be a nice UX if we also could load the promo code by hitting enter on the field. We have already a property on the component that does that: `onFieldSubmitted`. It's a callback that will be triggered when you hit enter, so you can just call the same function you call on the `Apply` button.

I'm approving, if you think it makes sense just let me know so I'll approve again.

**Comment by matibat** [comment]
Changes are carried to #1349

**Pull request closed without merging**